### DISCUSSION TITLE
Allow ipynb download of book chapter

### DISCUSSION THREAD

**Discussion by yogabonito** [quarto-dev/quarto-cli#11099] [Feature Requests]
### Description

I tried to create a book project with two output formats: html and ipynb. However, I don't see download link for ipynb in the preview and the Quarto prints a warning in the CLI: `WARN: The ipynb format is not supported by book projects`

Is there a plan to make the ipynb available as output format for book projects?

Thanks for considering my question and your great authoring software!

**Comment by cscheid** [comment]
Not immediately, no. The output formats for books that we plan to include are meant to be readable as an output format. HTML, epub, PDF, etc. The ipynb format _excels_ at being an intermediate representation for computation purposes, but doesn't currently offer a fantastic reading experience.

With that said, your individual chapters could have additional outputs as .ipynb for folks who eg might want to run the code in your chapters..

**Comment by yogabonito** [comment]
Thank you for the prompt reply, @cscheid!
I chose the book format for its cross-referencing capabilities. However, I would also like to include some exercises in the book. To allow my students to easily hop into the exercises, it would be really great to offer them the notebooks.

You wrote:

> With that said, your individual chapters could have additional outputs as .ipynb for folks who eg might want to run the code in your chapters..

Is that already somehow possible? I have now also tried to add

```
---
format:
  ipynb: default
---
```

to the top of a chapter, but it also doesn't work. Is there some hack to obtain a download link to an individual chapter as ipynb?

**Reply by mcanouil** [reply]
> I chose the book format for its cross-referencing capabilities

What do you mean? The cross-reference system works across all formats.


I don't think format linking is supported by Book project type.
Only `pdf` and `html` formats are supported, and only the formats defined in `_quarto.yml` are taken into account as far as I know.

As a workaround assuming a default Book project created with `quarto create project book`:
Make a pre-render script that
1. add the following to each chapters: (see `QUARTO_PROJECT_INPUT_FILES` https://quarto.org/docs/projects/scripts.html#pre-and-post-render)
  ```yaml
  ---
  format:
    html:
      other-links:
        - text: Notebook
          href: <FILENAME>.ipynb
  ---
  ```
2. Run `quarto convert <FILENAME>.qmd` on`QUARTO_PROJECT_INPUT_FILES`

Then you can `quarto render` (if you don't like the addition of the YAML header, you can make a post-render script)


It would be best if formats linking was working on Books though.

| `format-links` | `other-links` |
|--------|--------|
| <img width="1624" alt="image" src="https://github.com/user-attachments/assets/01a761f4-ed7f-4c27-a3e7-dd9945c537eb"> | <img width="1624" alt="image" src="https://github.com/user-attachments/assets/e707138e-e39f-46b1-bb2c-1304a2fdda20"> |

**Reply by yogabonito** [reply]
Thank you @mcanouil for taking the time to write this thorough reply! ❤️ I really appreciate it!

This workaround does work for me and I will keep using this solution. I would also appreciate the format linking feature for book projects, as it seems more elegant.

PS: What I meant by the cross-referencing capabilities of books is the following from quarto's [documentation](https://quarto.org/docs/books/): "HTML books use chapter numbers and therefore support Cross References between different chapters."
